src/graph/transform/to_bipartite.cc

```diff
@@ -0,0 +1,131 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file graph/transform/to_bipartite.cc
+ * \brief Convert a graph to a bipartite-structured graph.
+ */
+
+#include <dgl/base_heterograph.h>
+#include <dgl/transform.h>
+#include <dgl/array.h>
+#include <dgl/packed_func_ext.h>
+#include <dgl/immutable_graph.h>
+#include <dgl/runtime/registry.h>
+#include <dgl/runtime/container.h>
+#include <vector>
+#include <tuple>
+// TODO(BarclayII): currently ToBipartite depend on IdHashMap<IdType> implementation which
+// only works on CPU.  Should fix later to make it device agnostic.
+#include "../../array/cpu/array_utils.h"
+
+namespace dgl {
+
+using namespace dgl::runtime;
+using namespace dgl::aten;
+
+namespace transform {
+
+namespace {
+
+template<typename IdType>
+std::tuple<HeteroGraphPtr, std::vector<IdArray>, std::vector<IdArray>>
+ToBipartite(
+    HeteroGraphPtr graph,
+    const std::vector<IdArray> &rhs_nodes,
+    bool include_rhs_in_lhs) {
+  std::vector<IdHashMap<IdType>> rhs_node_mappings, lhs_node_mappings;
+  const int64_t num_etypes = graph->NumEdgeTypes();
+  const int64_t num_ntypes = graph->NumVertexTypes();
+  std::vector<EdgeArray> edge_arrays;
+
+  CHECK(rhs_nodes.size() == static_cast<size_t>(num_ntypes))
+    << "rhs_nodes not given for every node type";
+
+  for (const IdArray &nodes : rhs_nodes)
+    rhs_node_mappings.emplace_back(nodes);
```

Comment:
To make `rhs_node_mappings` local constant variables, I think this should work:
```c++
const std::vector<IdHashMap<IdType>> rhs_node_mappings(rhs_nodes.begin(), rhs_nodes.end());
```

Reply:
Thanks for bringing this syntax up.